fem/libceed/ceed-wrappers.cpp

```diff
@@ -0,0 +1,241 @@
+   : ess_tdofs(ess_tdofs_), P(P_)
+{
+   unconstrained_op = new UnconstrainedMFEMCeedOperator(oper);
+   Operator *rap = unconstrained_op->SetupRAP(P, P);
+   height = width = rap->Height();
+   bool own_rap = rap != unconstrained_op;
+   constrained_op = new ConstrainedOperator(rap, ess_tdofs, own_rap);
+}
+
+MFEMCeedOperator::MFEMCeedOperator(CeedOperator oper, const Operator *P_)
+   : MFEMCeedOperator(oper, Array<int>(), P_)
+{ }
+
+MFEMCeedOperator::~MFEMCeedOperator()
+{
+   delete constrained_op;
+   delete unconstrained_op;
+}
+
+void MFEMCeedOperator::Mult(const Vector& x, Vector& y) const
+{
+   constrained_op->Mult(x, y);
+}
+
+CeedOperator MFEMCeedOperator::GetCeedOperator() const
+{
+   return unconstrained_op->GetCeedOperator();
+}
+
+const Array<int> &MFEMCeedOperator::GetEssentialTrueDofs() const
+{
+   return ess_tdofs;
+}
+
+const Operator *MFEMCeedOperator::GetProlongation() const
+{
+   return P;
+}
+
+int MFEMCeedInterpolation::Initialize(
+   Ceed ceed, CeedBasis basisctof,
+   CeedElemRestriction erestrictu_coarse, CeedElemRestriction erestrictu_fine)
+{
+   int ierr = 0;
+
+   ierr = CeedInterpolationCreate(ceed, basisctof, erestrictu_coarse,
+                                  erestrictu_fine, &ceed_interp_); CeedChk(ierr);
+
+   ierr = CeedVectorCreate(ceed, height, &v_); CeedChk(ierr);
+   ierr = CeedVectorCreate(ceed, width, &u_); CeedChk(ierr);
+
+   return 0;
+}
+
+MFEMCeedInterpolation::MFEMCeedInterpolation(
+   Ceed ceed, CeedBasis basisctof,
+   CeedElemRestriction erestrictu_coarse,
+   CeedElemRestriction erestrictu_fine)
+{
+   int lo_nldofs, ho_nldofs;
+   CeedElemRestrictionGetLVectorSize(erestrictu_coarse, &lo_nldofs);
+   CeedElemRestrictionGetLVectorSize(erestrictu_fine, &ho_nldofs);
+   height = ho_nldofs;
+   width = lo_nldofs;
+   owns_basis_ = false;
+   Initialize(ceed, basisctof, erestrictu_coarse, erestrictu_fine);
+}
+
+MFEMCeedInterpolation::~MFEMCeedInterpolation()
+{
+   CeedVectorDestroy(&v_);
+   CeedVectorDestroy(&u_);
+   if (owns_basis_)
+   {
+      CeedBasisDestroy(&basisctof_);
+   }
+   CeedInterpolationDestroy(&ceed_interp_);
+}
+
+void MFEMCeedInterpolation::Mult(const mfem::Vector& x, mfem::Vector& y) const
+{
+   int ierr = 0;
+
+   const CeedScalar *x_ptr;
+   CeedScalar *y_ptr;
+   CeedMemType mem;
+   CeedGetPreferredMemType(internal::ceed, &mem);
+   if ( Device::Allows(Backend::CUDA) && mem==CEED_MEM_DEVICE )
+   {
+      x_ptr = x.Read();
+      y_ptr = y.ReadWrite();
+   }
+   else
+   {
+      x_ptr = x.HostRead();
+      y_ptr = y.HostReadWrite();
+      mem = CEED_MEM_HOST;
+   }
+
+   ierr += CeedVectorSetArray(u_, mem, CEED_USE_POINTER,
+                              const_cast<CeedScalar*>(x_ptr));
+   ierr += CeedVectorSetArray(v_, mem, CEED_USE_POINTER, y_ptr);
+
+   ierr += CeedInterpolationInterpolate(ceed_interp_, u_, v_);
+
+   ierr += CeedVectorTakeArray(u_, mem, const_cast<CeedScalar**>(&x_ptr));
+   ierr += CeedVectorTakeArray(v_, mem, &y_ptr);
```

Comment:
It would be nice here also to use `CeedAddMult`

Reply:
Fixed, but my solution is a little bit of a hack. Let me know if you have a better idea.

Reply:
The pointwise mult of the vector makes this one a little harder to fix than the other. I think you're right that we should do something like this to avoid duplicating code, but I'll have to think about exactly how to restructure.

Reply:
I didn't use `CeedAddMult` but I think I've improved the organization / modularity.